src/api/action/search.ts

```diff
@@ -0,0 +1,122 @@
+
+import {
+    type HonoContextVariables,
+    fedify
+} from '../../app';
+
+interface ProfileSearchResult {
+    actor: any;
+    handle: string;
+    followerCount: number;
+    isFollowing: boolean;
+}
+
+interface SearchResults {
+    profiles: ProfileSearchResult[];
+}
+
+// @<username>@<domain>.<tld>
+const HANDLE_REGEX = /^@([\w-]+)@([\w-]+\.[\w.-]+)$/;
+
+// http(s)://...
+const URI_REGEX = /^https?:\/\/[^\s/$.?#].[^\s]*$/;
+
+export async function searchAction(
+    ctx: Context<{ Variables: HonoContextVariables }>,
+) {
+    const db = ctx.get('db');
+    const apCtx = fedify.createContext(ctx.req.raw as Request, {
+        db,
+        globaldb: ctx.get('globaldb'),
+    });
+
+    // Parse "query" from query parameters
+    const query = ctx.req.query('query') || '';
+
+    // Init search results - At the moment we only support searching for an actor (ui calls them profiles)
+    const results: SearchResults = {
+        profiles: [],
+    };
+
+    // If the query is not a handle or URI, return early
+    if (HANDLE_REGEX.test(query) === false && URI_REGEX.test(query) === false) {
+        console.log(`Invalid query: ${query}`);
+
+        return new Response(JSON.stringify(results), {
+            headers: {
+                'Content-Type': 'application/json',
+            },
+            status: 200,
+        });
+    }
+
+    // Lookup actor by handle or url
+    try {
+        const actor = await apCtx.lookupObject(query);
+
+        if (actor && isActor(actor)) {
+            const result: ProfileSearchResult = {
+                actor: {},
+                handle: '',
+                followerCount: 0,
+                isFollowing: false,
+            };
+
+            // Retrieve actor data
+            result.actor = await actor.toJsonLd();
+
+            // Sanitize actor data
+            const sanitizeHtmlConfig = {
+                allowedTags: ['a', 'p', 'img', 'br', 'strong', 'em', 'span'],
+                allowedAttributes: {
+                    a: ['href'],
+                    img: ['src'],
+                }
+            };
+
+            result.actor.summary = sanitizeHtml(result.actor.summary, sanitizeHtmlConfig);
+
+            if (result.actor.attachment) {
+                result.actor.attachment = result.actor.attachment.map((attachment: any) => {
+                    if (attachment.type === 'PropertyValue') {
+                        attachment.value = sanitizeHtml(attachment.value, sanitizeHtmlConfig);
+                    }
+
+                    return attachment;
+                });
+            }
+
+            // Compute the full handle for the actor
+            result.handle = `@${actor.preferredUsername}@${actor.id!.host}`;
+
+            // Retrieve follower count for the actor
+            const followers = await ky
+                .get(actor.followersId!.href)
+                .json<{ totalItems: number }>();
+
+            result.followerCount = followers.totalItems;
```

Comment:
If you want to get a collection of followers, you can easily do so with `await actor.getFollowers()`:

```suggestion
            result.followerCount = (await actor.getFollowers()).totalItems;
```

I guess you're maintaining the cache for ActivityStreams objects, it would be more efficient if you implement a `DocumentLoader` so that it looks for the cache first before actually fetching an ActivityStreams object through the network. 🤔

Reply:
Aha, yes this is what I was looking for, thanks!

> I guess you're maintaining the cache for ActivityStreams objects, it would be more efficient if you implement a DocumentLoader so that it looks for the cache first before actually fetching an ActivityStreams object through the network. 🤔

Yep, we already do this elsewhere / have this functionality 👍

Reply:
Was thinking about this a little more @dahlia 

Do you think it would be useful to have this functionality within Fedify itself, so you could just do:

```js
await actor.getFollowersCount()
```